### ISSUE TITLE
Inputs-from-file argument not recognized

### ISSUE THREAD

**Issue by pbilling** [googlegenomics/dockerflow#20]
Hi Jonathan,

I'm trying to use the "--inputs-from-file" command-line argument to pass a text file, with a list of input files, to dockerflow. I've tried to emulate your "Parallel workflows" example as close as possible, but have not been able to get it to work.

I am using your parallel-graph.yaml workflow and specifying inputs from a separate file. I changed the names of the input files in "many_files.txt" to verify whether the command line argument was overriding the defaults in the workflow file, and found they were not.

    $ dockerflow --project=gbsc-gcp-project-mvp \
    --workflow-file=src/test/resources/parallel-graph.yaml \
    --workspace=gs://gbsc-gcp-project-mvp-group/test/dockerflow_test \
    --inputs-from-file=stepOne.inputFile=many_files.txt \
    --test \
    --runner=DirectPipelineRunner

The content of "many_files.txt" is as follows:

    REQUIRED/input-one-file-input.txt
    REQUIRED/input-two-file-input.txt

I also tried specifying the full gs://path/to/file in "many_files.txt", to no effect. It was unclear to me whether the file should be stored locally or in the cloud workspace, so I tried both, to no avail.

When I removed the arguments from the workflow file, and only specified on the command-line, I got the the following NullPointer Exception:

    INFO: Resolving paths vs gs://gbsc-gcp-project-mvp-group/test/dockerflow_test/stepOne/
    Exception in thread "main" com.google.cloud.dataflow.sdk.Pipeline$PipelineExecutionException: java.lang.NullPointerException: Null value for path inputFile in task stepOne
        at com.google.cloud.dataflow.sdk.Pipeline.run(Pipeline.java:187)
        at com.google.cloud.genomics.dockerflow.Dockerflow.main(Dockerflow.java:172)
    Caused by: java.lang.NullPointerException: Null value for path inputFile in task stepOne
        at com.google.cloud.genomics.dockerflow.task.Task.resolvePaths(Task.java:426)
        at com.google.cloud.genomics.dockerflow.task.Task.applyArgs(Task.java:254)
        at com.google.cloud.genomics.dockerflow.transform.DockerDo$ScatterTasks.processElement(DockerDo.java:318)

It seems like I am specifying the "--input-from-file" in a way that it is not being recognized, but I can't figure why not. 

Any guidance is greatly appreciated!

Best regards,
Paul

**Comment by jbingham** [comment]
tl;dr: The --inputs-from-file option only supports input parameters that are environment variables, not input and output file parameters. This is as designed, though unfortunately not documented. Please file a new issue if you'd like to see support added.

What's the ultimate goal that you have? You have multiple files listed inside of your file in GCS. What do you want to do with each of those files? Are you trying to download them all to the VM's local disk? Why not download them using a wildcard? Or create a separate inputParameter entry for each one?

More notes:

I've confirmed that if you set the value of stepOne.message using --inputs-from-file=message=file.txt, it works as expected. If you set the value of stepOne.inputFile, it does not. 

First, confirm that the example works:
> dockerflow --project=my-project --workspace=gs://my-bucket/my-path --workflow-file=src/test/resources/parallel-graph.yaml --test --runner=DirectPipelineRunner --inputs=BASE_DIR=gs://my-bucket/my-path

Then, create a local file, file.txt:
> dockerflow --project=my-project --workspace=gs://my-bucket/my-path --workflow-file=src/test/resources/parallel-graph.yaml --test --runner=DirectPipelineRunner --inputs=BASE_DIR=gs://my-bucket/my-path --inputs-from-file=stepOne.message=file.txt

This works. But this doesn't:
> dockerflow --project=my-project --workspace=gs://my-bucket/my-path --workflow-file=src/test/resources/parallel-graph.yaml --test --runner=DirectPipelineRunner --inputs=BASE_DIR=gs://my-bucket/my-path --inputs-from-file=stepOne.inputFile=file.txt

This is as designed, since input files require some specific fancy handling to ensure they get copied to local disk correctly.

**Comment by pbilling** [comment]
Hi Jonathan,

OK, thanks for clarifying this. My use case is that I have many thousands of input files, and for each one, I want to run a separate instance of the same workflow, in parallel.

Knowing that the --inputs-from-file option only supports environment variables, I think I'll try passing each file path as a variable and then add a command to my docker script to download the file to the VM.

Best regards,
Paul

**Comment by pbilling** [comment]
Follow-up question: can I scatter by inputs specified in a file? For instance, in the test case you described, could I specify a different stepOne.message on each line of a file, and have dockerflow scatter by those messages? Or will everything in the file be read as a single stepOne.message value?

**Comment by pbilling** [comment]
Update: I tried running the first example with the --inputs-from-file=stepOne.message=file.txt argument, but changing the text in file.txt. The task completed successfully, but still used the default value of "hello" instead of the one I provided in file.txt.

**Comment by jbingham** [comment]
Thanks for the context on what you're trying to accomplish.

Yes, the scattering functionality is intended to allow providing a
multi-line file with a list of values. For example, if you look in
examples/gatk, that code scatters by calling intervals, and spins up around
50 concurrent VMs.

I'm running a test now to confirm that the input-from-file parameter is
working correctly.

**Comment by jbingham** [comment]
In my test, the values from the file passed to --input-from-file ended up in the output files as expected. 

dockerflow --project=my-project --workspace=gs://my-bucket/path --workflow-file=src/test/resources/parallel-graph.yaml --inputs=BASE_DIR=gs://my-bucket/path --inputs-from-file=stepOne.message=file.txt

See stepOne/1/output-one.txt and stepOne/2/output-one.txt. 

When running with --test and --runner=DirectPipelineRunner, you can confirm that the right value is passed to the tasks by looking at the json output for  "pipelineArgs". There, the value of "message" should be the contents of your file.

The GATK example is probably the best place for what you're looking for, even though it's very long and you'll have to ignore most of it. Open the yaml file and search for "scatter" to find where a file is loaded with the values to scatter by.

**Comment by pbilling** [comment]
Thanks for the tip! Using --test and --runner=DirectPipelineRunner to validate pipeline arguments is very handy. And I've been using the GATK example for reference; it's long but I've found a lot of useful info in it.

I just re-visited this and found the issue to be a typo in line 62 of /main/java/com/google/cloud/genomics/dockerflow/DockerflowConstants.java: 
```
String INPUTS_FROM_FILE = "inputs-fron-file";
```
Now it seems to work fine. Thanks for all the help!

**Comment by jbingham** [comment]
Ah, thanks for catching the typo. I've created a PR to fix this.
